### ISSUE TITLE
Switch to new Java classes to manage dates

### ISSUE THREAD

**Issue by lvca** [ArcadeData/arcadedb#608]
### Discussed in https://github.com/ArcadeData/arcadedb/discussions/605

ArcadeDB uses `java.util.Date` to manage dates. Recent versions of Java provide more powerful classes to manage dates (Timestamp, LocalDatetime, etc) that also support granularity higher than millisecond.

**Comment by lvca** [comment]
In the new branch, we've introduced new methods in the Document:

```java

  /**
   * Returns a java.util.Calendar object from a datetime property.
   *
   * @since 23.1.1
   */
  Calendar getCalendar(String propertyName);

  /**
   * Returns a java.time.LocalDate object from a date property.
   *
   * @since 23.1.1
   */
  LocalDate getLocalDate(String propertyName);

  /**
   * Returns a java.time.LocalDateTime object from a datetime property.
   *
   * @since 23.1.1
   */
  LocalDateTime getLocalDateTime(String propertyName);

  /**
   * Returns a java.time.ZonedDateTime object from a datetime property.
   *
   * @since 23.1.1
   */
  ZonedDateTime getZonedDateTime(String propertyName);

  /**
   * Returns a java.time.Instant object from a datetime property.
   *
   * @since 23.1.1
   */
  Instant getInstant(String propertyName);
```

**Comment by lvca** [comment]
Also, there are 3 new settings:

```
  DATE_IMPLEMENTATION("arcadedb.dateImplementation",
      "Default date implementation to use on deserialization. By default java.util.Date is used, but the following are supported: java.util.Calendar, java.time.LocalDate",
      Class.class, java.util.Date.class),

  DATE_TIME_IMPLEMENTATION("arcadedb.dateTimeImplementation",
      "Default datetime implementation to use on deserialization. By default java.util.Date is used, but the following are supported: java.util.Calendar, java.time.LocalDateTime, java.time.ZonedDateTime",
      Class.class, java.util.Date.class),

  DATE_TIME_PRECISION("arcadedb.dateTimePrecision",
      "Precision to use with Datetime types on deserialization. Check the level of precision supported by your OS/JVM. By default is 'millisecond'. Supported are: 'millisecond', 'microsecond' and 'nanosecond'",
      String.class, "millisecond"),
```

You can use Java API to modify the settings, but better saving those settings in the database with the `ALTER DATABASE` SQL statement:

```java
    database.command("sql", "alter database `arcadedb.dateTimeImplementation` `java.time.LocalDateTime`");
    database.command("sql", "alter database `arcadedb.dateImplementation` `java.time.LocalDate`");
    database.command("sql", "alter database `arcadedb.dateTimePrecision` nanosecond");
```

**Comment by lvca** [comment]
Once you set the precision (datetime only), ArcadeDB will use that setting to save the timestamp that supports that precision. With nanoseconds, it's a really long number, so bear this in mind. Longer number = more space used on disk and in RAM. 

Since the datetime precision is a global setting (or database setting), the precision is not saved with each datetime property. This means if you change the precision after some of the date-time properties have been stored in your records, all your datetimes must be updated in the database.

For example, if you used nanoseconds:

```sql
ALTER DATABASE `arcadedb.dateTimePrecision` nanosecond
```

And you had the property "signedOn" as DATETIME on a "Profile" vertex type, after a while, you understand you don't need that level of precision, you can change it back to milliseconds.

```sql
ALTER DATABASE `arcadedb.dateTimePrecision` millisecond
```

Right after having changed the setting, you should run this query:

```sql
UPDATE Profile SET signedOn = signedOn / 1000000
```

Since the difference between nanoseconds (10E9) and milliseconds (10E3) is 10E6 = 1 million. The query above updates the date by dividing 1 million from the date (that is stored as a variable length number).

**Comment by lvca** [comment]
In the future, we could support custom precision at the property level with something like:

```sql
ALTER PROPERTY `EventLog.date` SET precision = 'microsecond'
```

**Comment by lvca** [comment]
100% implemented. Docs are still pending.

**Comment by vic0824** [comment]
So there is still only one Datetime data type, but now there are new settings to specify the underlying implementation and the precision.
When you say "We will eventually switch the default implementation to the new API", does it mean that you will eliminate the two  *Implementation settings and keep only the precision setting? 
I was expecting this enhancement to be implemented by introducing new data types, but I can see your point about wasting disk space and RAM if you always store a LocalDateTime with maximum precision even when you don't need it.

I wonder what traditional relational DBMS do when they use the sql TIMESTAMP data type: do they always store it with nanoseconds precision or do they infer the precision from the literal values provided by the user? I guess it depends on the db internals, i.e. whether they store a fixed-size record in a page or not.
This sparked my curiosity so I had a look at the `java.sql.Timestamp` implementation (which has its well-known problems).

The apidoc says:

```
The precision of a Timestamp object is calculated to be either:

19 , which is the number of characters in yyyy-mm-dd hh:mm:ss
20 + s , which is the number of characters in the yyyy-mm-dd hh:mm:ss.[fff...] and s represents the scale of the given Timestamp, its fractional seconds precision.
```

but the source code of the `valueOf(LocalDateTime dateTime)` method (which is what I used in my project that uses Hsqldb) is:

```
    public static Timestamp valueOf(LocalDateTime dateTime) {
        return new Timestamp(dateTime.getYear() - 1900,
                             dateTime.getMonthValue() - 1,
                             dateTime.getDayOfMonth(),
                             dateTime.getHour(),
                             dateTime.getMinute(),
                             dateTime.getSecond(),
                             dateTime.getNano());
    }
```

so it's only possible to specify the accuracy when using the `valueOf(String s)` method, whereas the `valueOf(LocalDateTime dateTime)` method by default stores also nanoseconds.
But then again, java.sqlTimestamp holds a `private int nanos;` inside so it always uses the maximum space.

Perhaps an alternative solution would be to define different types, each one bound to a pre-defined precision, for example:

- Datetime: it has a resolution of 1 second 
- DatetimeMillis: it has a resolution of 1 millisecond
- DatetimeMicros: it has a resolution of 1 microsecond
- DatetimeNanos: it has a resolution of 1 nanosecond

This would allow to engine to allocate only the minimum required space for each data type, and if it's implemented using the new API, you could even create more types for cases when you only need the date and not the time, thus further optimizing the resources.
In this way, there would be no need for an additional setting to configure the precision, but it would lead to a proliferation of types, so there is no clear winner, it is more a matter of personal preference.

If you go along the route of the precision setting, you could even allow setting it to "seconds", "minutes", "hours", "days", to optimize the space allocated even further.

**Comment by lvca** [comment]
The `getNano()` in LocalDateTime returns the underlying `nano` integer that stores the fractional part of the second. If you have millisecond accuracy, you'll have `333000000`, with microseconds it would be `333333000` and with nanos would be `333333333`. In this way you can use one variable to store millis, micros and nanos. The same have done with `java.sql.Timestamp`.

With ArcadeDB dates are always numbers. Since ArcadeDB already compresses numbers by using only the bytes required (it's a variable number) it's already ok for storing any big number. So it was much easier to convert all the date-time types into numbers. The missing information is the level of accuracy. it would be much easier to store an additional property that tells the accuracy of the date, like `N:3473673647346` that tells us it's in (N) nanoseconds, but it would break the current implementation. Not a big deal if we'd use new datatypes as you suggested.

Also, you can have nanosecond accuracy, but you want to work with ZonedDateTime instead of LocalDateTime. Or also Instant. In Java now there are multiple implementations for dates, so we'd need to let ArcadeDB which implementation the user is comfortable with and we get and return that class every time without letting the user convert it into its date class.

So I like your idea about using new types for dates. Since >99% of people will be ok with milliseconds precision and we don't want to break things unless it's fundamental to the progress, what about introducing 2 new types?
- `DATETIME_MICROS` for microseconds
- `DATETIME_NANOS` for nanoseconds 
- `DATETIME_SECONDS` for just seconds. I believe tons of people should use this when they manage date time types, but we can't make it the default without breaking existent databases and code

In this way, the current DATETIME would work with milliseconds without breaking the compatibility with the past.

Also, I could get rid of the new `precision` attribute in Properties. You don't need to set the precision anymore, just different date types. 

Is anybody cool with that? WDYT @gramian? In this case I need a few more days to finalize such changes before 23.1.1 and docs updated.

**Comment by vic0824** [comment]
Fine with me.

**Comment by gramian** [comment]
Fine with me too. Please cherry pick my recent docs PR.

**Comment by lvca** [comment]
Ok, I've already started this update. The nice thing about having new `datetime_*` is that only the real precision is stored, with zero waste of space. I'm leaving the `.precision(<precision>)` sql method that is still the best way to massive update all the dates to a different precision. Super useful if you want to truncate precision with just one command.

The new `precision` attribute in Property is gone, restoring the old Property class. Also, the global setting about precision doesn't make any sense anymore. It's all about each value and the precision used.

Thanks, @vic0824 this was a very good idea! Glad you said something before releasing v23.1.1.

**Comment by lvca** [comment]
Finalized the new changes.

Introduced new types:

- `DATETIME_SECOND` for precision up to the second
- `DATETIME_MICROS` for precision up to the microsecond
- `DATETIME_NANOS` for precision up to the nanosecond

This means the old `DATETIME` type still works with millisecond precision and it's still compatible with Java types `java.util.Date` and `java.util.Calendar`.

**Comment by lvca** [comment]
Docs updated, I'm closing this issue just in time for the release 23.1.1 scheduled for tomorrow.

**Comment by vic0824** [comment]
How do I set the DateTime implementation via the Java API?
`db.getConfiguration().setValue(GlobalConfiguration.DATE_IMPLEMENTATION, java.time.LocalDateTime.class)`
or `db.getConfiguration().setValue(GlobalConfiguration.DATE_IMPLEMENTATION, "java.time.LocalDateTime")`

**Comment by vic0824** [comment]
I'm starting to use the new types, but I don't understand how to get a LocalDateTime object from the output of the getProperty method.
If I create a property of type Type.DATETIME_MICROS, and I insert the following LocalDateTime instance:
`LocalDateTime vstart = LocalDateTime.parse("20220319_002905.423534", DateTimeFormatter.ofPattern("yyyyMMdd'_'HHmmss.SSSSSS"));`
when I do a select, getProperty returns this value: `Mon Dec 10 05:50:23 CET 54181`
I attach the test method, and this is the output I get:
```
INSERT INTO Order SET id = ?, processor = ?, vstart = ?, vstop = ?, pstart = ?, pstop = ?, status = ?, node = ?
parameters: 1, SIR1LRM-7.1, 2022-03-19T00:29:05.423534, 2022-03-19T00:34:19.571172, 2023-01-18T08:30:00.123, 2023-01-18T08:32:00.456, PENDING, cs2minipds
result = {"id":1,"processor":"SIR1LRM-7.1","vstart":1647649745423534,"vstop":1647650059571172,"pstart":1674030600123,"pstop":1674030720456,"status":"PENDING","node":"cs2minipds","@cat":"d","@type":"Order","@rid":"#1:0"}
select from Order
id = 1, processor = SIR1LRM-7.1, vstart = Mon Dec 10 05:50:23 CET 54181, vstop = Thu Dec 13 21:06:11 CET 54181, pstart = Wed Jan 18 09:30:00 CET 2023, pstop = Wed Jan 18 09:32:00 CET 2023, status = PENDING, node = cs2minipds
```
If I try to cast the return value of getProperty to LocalDateTime I get a class cast exception: `class java.util.Date cannot be cast to class java.time.LocalDateTime`, so it returns a Date even if I have set the arcadedb.dateTimeImplementation property as you suggest.


[testDateTimeMicros.txt](https://github.com/ArcadeData/arcadedb/files/10448512/testDateTimeMicros.txt)

**Comment by lvca** [comment]
Have you tried to get the object with `getLocalDateTime()`

**Comment by vic0824** [comment]
It still does not work: the last 5 digits of the fractional seconds part are interpreted as seconds and milliseconds of a Date object.
I think the problem is how the values are inserted, not how they are retrieved.
It seems that the values are inserted as Date objects even if the property is defined as DATETIME_MICROS.
I have also tried to set arcadedb.dateTimeFormat to "yyyy-MM-dd'T'HH:mm:ss.SSSSSS", but it still does not work.
Can you provide an example of how to insert a LocalDateTime object into a DATETIME_MICROS property?

**Comment by lvca** [comment]
Check this test case: https://github.com/ArcadeData/arcadedb/blob/a3f1b8430407426c6d6492092d32a21d20941d20/engine/src/test/java/com/arcadedb/DateTest.java#L179

**Comment by vic0824** [comment]
Sorry, I should have been more specific: your test case proves very clearly that the Java Document API works as expected, but it does not prove that the SQL API works as expected.
I think my SQL test case shows that a simple insert does not work as expected: I insert a LocalDateTime value with microsecond precision into a DATETIME_MICROS property, and the engine saves it as a Date with completely different year, month, day, hours, minutes, seconds. There is no way to retrieve the value that was inserted, because it has been incorrectly encoded when it was written into the db. In particular, it seems that it has taken the n least significant bytes of the LocalDateTime and interpreted them as a Date.
If you prefer, I can open a dedicated issue.

**Comment by lvca** [comment]
I agree we lacked some additional test cases by using SQL. Working on this today.

**Comment by vic0824** [comment]
I have found another issue: `MutableDocument.set(final String name, Object value)` throws `Type 'class java.time.LocalDateTime' not supported for JSONObject` if the Document contains a property of type Type.DATETIME_MICROS.

**Comment by lvca** [comment]
Could you please provide a test case for your last comment or a way to reproduce it?

Look at this: https://github.com/ArcadeData/arcadedb/blob/a3f1b8430407426c6d6492092d32a21d20941d20/engine/src/test/java/com/arcadedb/DateTest.java#L206, it's storing a LocalDateTime of type Type.DATETIME_MICROS.

**Comment by lvca** [comment]
Reproduced the issue with SQL.

**Comment by vic0824** [comment]
Never mind. I must have been doing something wrong, because when I wrote a test case, the problem did not occur.

**Comment by lvca** [comment]
Tested and works with SQL too. My test failed because I didn't set the LocalDateTime as implementation. This test passes:

```java
@Test
  public void testSQL() throws ClassNotFoundException {
    final LocalDateTime localDateTime = LocalDateTime.now();

    database.command("sql", "alter database dateTimeImplementation `java.time.LocalDateTime`");
    try {
      database.begin();
      ResultSet result = database.command("sql", "insert into ConversionTest set datetime_second = ?", localDateTime);
      Assertions.assertTrue(result.hasNext());
      Assertions.assertEquals(localDateTime.truncatedTo(ChronoUnit.SECONDS), result.next().toElement().get("datetime_second"));

      result = database.command("sql", "insert into ConversionTest set datetime_millis = ?", localDateTime);
      Assertions.assertTrue(result.hasNext());
      Assertions.assertEquals(localDateTime.truncatedTo(ChronoUnit.MILLIS), result.next().toElement().get("datetime_millis"));

      result = database.command("sql", "insert into ConversionTest set datetime_micros = ?", localDateTime);
      Assertions.assertTrue(result.hasNext());
      Assertions.assertEquals(localDateTime.truncatedTo(ChronoUnit.MICROS), result.next().toElement().get("datetime_micros"));

      result = database.command("sql", "insert into ConversionTest set datetime_nanos = ?", localDateTime);
      Assertions.assertTrue(result.hasNext());
      Assertions.assertEquals(localDateTime.truncatedTo(ChronoUnit.NANOS), result.next().toElement().get("datetime_nanos"));

      database.commit();
    } finally {
      database.command("sql", "alter database dateTimeImplementation `java.util.Date`");
    }
  }
```

Are you still able to reproduce the issue with SQL?

**Comment by vic0824** [comment]
I have found the reason.
In your test case, you are using the same Database object that you created with the Factory, whereas in my test case I'm using a Database object obtained from the embedded server.
I execute the alter database command on the Database object created with the Factory, but then I start the server with the default ContextConfiguration, which is created with the default GlobalConfiguration values, which apparently take precedence over the database settings saved in the database.
Sure enough, if I use 
```
GlobalConfiguration.DATE_TIME_IMPLEMENTATION.setValue(java.time.LocalDateTime.class);
ArcadeDBServer arcadeDBServer = new ArcadeDBServer(new ContextConfiguration());
```
it works.
On the other hand, if I use:
```
Database database = arcadeDBServer.getDatabase("test");
database.getConfiguration().setValue(GlobalConfiguration.DATE_IMPLEMENTATION, java.time.LocalDateTime.class);
```
it does not work.
Is this the expected behaviour?
I admit I might be victim of my own enhancement request for database vs server settings, but according to the manual 

> You can access to the database configuration with database.getConfiguration() to read and write per database settings

so I had assumed that I could save the setting in the db using database.getConfiguration().setValue(GlobalConfiguration.DATE_IMPLEMENTATION, java.time.LocalDateTime.class);

**Comment by vic0824** [comment]
There is still one small issue: both `Result.toJSON()` and `Result.toElement().toJSON()` throw `Type 'class java.time.LocalDateTime' not supported for JSONObject`

**Comment by lvca** [comment]
I can confirm that the settings stored in the database should always take precedence, so it's a bug.

**Comment by lvca** [comment]
The reason why this doesn't work:

```java
Database database = arcadeDBServer.getDatabase("test");
database.getConfiguration().setValue(GlobalConfiguration.DATE_IMPLEMENTATION, java.time.LocalDateTime.class);
```

Is because the setting is saved in the configuration, but the binary serializer is already set with the loaded setting. The right way to change this configuration and have effect immediately (otherwise you have to re-open the database) is to execute an `alter database` SQL command.

From the test case and nby playing with a server, I see the settings are correctly stored and used in the running database. How can I reproduce your exact case? Could you please provide a piece of code where you start the server as embedded and the LocalDateTime is not taken?

**Comment by vic0824** [comment]
To which test case are you referring? Executing alter database before starting the server or using database.getConfiguratation.setValue after starting the server?
I attach both:

- testDateTimeMicros1: execute alter database and then start the server
- testDateTimeMicros2: start the server and then execute arcadeDBServer.getDatabase("test").getConfiguration().setValue

What about the exceptions thrown by the toJSON methods? Is it a bug or I should use a different method to get a JSON representation of the inserted record?

[testDateTimeMicros1.zip](https://github.com/ArcadeData/arcadedb/files/10473970/testDateTimeMicros1.zip)
[testDateTimeMicros2.zip](https://github.com/ArcadeData/arcadedb/files/10473971/testDateTimeMicros2.zip)

**Comment by lvca** [comment]
Thanks @vic0824 for the test cases. I run the 1st one and the output seems correct:

```
/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/java -ea -Dorg.slf4j.simpleLogger.defaultLogLevel=warn -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn -Dpolyglot.engine.WarnInterpreterOnly=false -Didea.test.cyclic.buffer.size=1048576 -javaagent:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=59996:/Applications/IntelliJ IDEA CE.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /Users/luca/.m2/repository/org/junit/platform/junit-platform-launcher/1.9.2/junit-platform-launcher-1.9.2.jar:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit5-rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit-rt.jar:/Users/luca/Documents/GitHub/arcadedb/server/target/test-classes:/Users/luca/Documents/GitHub/arcadedb/server/target/classes:/Users/luca/Documents/GitHub/arcadedb/engine/target/classes:/Users/luca/.m2/repository/com/conversantmedia/disruptor/1.2.21/disruptor-1.2.21.jar:/Users/luca/.m2/repository/org/lz4/lz4-java/1.8.0/lz4-java-1.8.0.jar:/Users/luca/.m2/repository/org/apache/lucene/lucene-analyzers-common/8.11.2/lucene-analyzers-common-8.11.2.jar:/Users/luca/.m2/repository/org/apache/lucene/lucene-core/8.11.2/lucene-core-8.11.2.jar:/Users/luca/.m2/repository/org/apache/lucene/lucene-queryparser/8.11.2/lucene-queryparser-8.11.2.jar:/Users/luca/.m2/repository/org/apache/lucene/lucene-queries/8.11.2/lucene-queries-8.11.2.jar:/Users/luca/.m2/repository/org/apache/lucene/lucene-sandbox/8.11.2/lucene-sandbox-8.11.2.jar:/Users/luca/.m2/repository/com/google/code/gson/gson/2.10.1/gson-2.10.1.jar:/Users/luca/.m2/repository/org/graalvm/sdk/graal-sdk/22.3.0/graal-sdk-22.3.0.jar:/Users/luca/.m2/repository/org/graalvm/js/js/22.3.0/js-22.3.0.jar:/Users/luca/.m2/repository/org/graalvm/regex/regex/22.3.0/regex-22.3.0.jar:/Users/luca/.m2/repository/org/graalvm/truffle/truffle-api/22.3.0/truffle-api-22.3.0.jar:/Users/luca/.m2/repository/com/ibm/icu/icu4j/71.1/icu4j-71.1.jar:/Users/luca/Documents/GitHub/arcadedb/network/target/classes:/Users/luca/Documents/GitHub/arcadedb/integration/target/classes:/Users/luca/.m2/repository/com/univocity/univocity-parsers/2.9.1/univocity-parsers-2.9.1.jar:/Users/luca/.m2/repository/io/undertow/undertow-core/2.3.3.Final/undertow-core-2.3.3.Final.jar:/Users/luca/.m2/repository/org/jboss/logging/jboss-logging/3.4.3.Final/jboss-logging-3.4.3.Final.jar:/Users/luca/.m2/repository/org/jboss/xnio/xnio-api/3.8.8.Final/xnio-api-3.8.8.Final.jar:/Users/luca/.m2/repository/org/wildfly/common/wildfly-common/1.5.4.Final/wildfly-common-1.5.4.Final.jar:/Users/luca/.m2/repository/org/wildfly/client/wildfly-client-config/1.0.1.Final/wildfly-client-config-1.0.1.Final.jar:/Users/luca/.m2/repository/org/jboss/xnio/xnio-nio/3.8.8.Final/xnio-nio-3.8.8.Final.jar:/Users/luca/.m2/repository/org/jboss/threads/jboss-threads/3.5.0.Final/jboss-threads-3.5.0.Final.jar:/Users/luca/.m2/repository/org/slf4j/slf4j-api/2.0.6/slf4j-api-2.0.6.jar:/Users/luca/.m2/repository/org/slf4j/slf4j-jdk14/2.0.6/slf4j-jdk14-2.0.6.jar:/Users/luca/.m2/repository/io/dropwizard/metrics/metrics-core/4.2.15/metrics-core-4.2.15.jar:/Users/luca/.m2/repository/io/dropwizard/metrics/metrics-jmx/4.2.15/metrics-jmx-4.2.15.jar:/Users/luca/.m2/repository/org/junit/jupiter/junit-jupiter/5.9.2/junit-jupiter-5.9.2.jar:/Users/luca/.m2/repository/org/junit/jupiter/junit-jupiter-api/5.9.2/junit-jupiter-api-5.9.2.jar:/Users/luca/.m2/repository/org/opentest4j/opentest4j/1.2.0/opentest4j-1.2.0.jar:/Users/luca/.m2/repository/org/junit/platform/junit-platform-commons/1.9.2/junit-platform-commons-1.9.2.jar:/Users/luca/.m2/repository/org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.jar:/Users/luca/.m2/repository/org/junit/jupiter/junit-jupiter-params/5.9.2/junit-jupiter-params-5.9.2.jar:/Users/luca/.m2/repository/org/junit/jupiter/junit-jupiter-engine/5.9.2/junit-jupiter-engine-5.9.2.jar:/Users/luca/.m2/repository/org/junit/platform/junit-platform-engine/1.9.2/junit-platform-engine-1.9.2.jar:/Users/luca/.m2/repository/org/assertj/assertj-core/3.24.2/assertj-core-3.24.2.jar:/Users/luca/.m2/repository/net/bytebuddy/byte-buddy/1.12.21/byte-buddy-1.12.21.jar:/Users/luca/.m2/repository/org/mockito/mockito-core/4.11.0/mockito-core-4.11.0.jar:/Users/luca/.m2/repository/net/bytebuddy/byte-buddy-agent/1.12.19/byte-buddy-agent-1.12.19.jar:/Users/luca/.m2/repository/org/objenesis/objenesis/3.3/objenesis-3.3.jar com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit5 com.arcadedb.server.SelectOrderRIDTest,testDateTimeMicros1

INSERT INTO Order SET vstart = ?
parameters: 2023-01-22T19:47:00.511144
resultSet.hasNex() = true
select from Order
vstart = 2023-01-22T19:47:00.511144
```

**Comment by lvca** [comment]
Thanks to your test cases 1 I was able to reproduce both issues with the settting of datetime implementation at database load and the JSON error. Working on it.

**Comment by lvca** [comment]
Ok, now the problem is fixed. I've slightly modified your test case to be part of our test cases suite to avoid regressions. Thanks for the test cases!

By the way, the 2nd test case wouldn't work because the right way is to execute an alter database, not changing the configuration (unless you do that and then reopen the database).

**Comment by vic0824** [comment]
Sorry to come back again to this, but I believe `BinaryComparator` should be updated to handle the new DATETIME_* types and in general to use the new LocalDateTime implementation.

I'm getting sporadic exceptions like this:
```
java.lang.IllegalArgumentException: Comparison between type 20 and 20 not supported
        at com.arcadedb.serializer.BinaryComparator.compare(BinaryComparator.java:766)
```
when executing SQL statements on a DATETIME_MICROS type.
I'm currently unable to create a simple test case to reproduce the problem: it always happens in my code, but it does not happen in the simplified test case I've written, and I don't understand why, I'll keep working on it.

**Comment by lvca** [comment]
@vic0824 I think you're not using the latest snapshot. I can't find any code at BinaryComparator.java:766. Please let me know if the same issue happens with the latest branch.

**Comment by vic0824** [comment]
Pulled latest branch, now all my custom handlers don't work anymore, I get a 404 and no log messages from ArcadeDB or Undertow, even with logging level set to FINE.
Has something changed in the way the handlers are passed to Undertow?

**Comment by vic0824** [comment]
I have finally managed to create a test case that reproduces the problem, using v23.1.2-SNAPSHOT (build 45102bde1d01085636101f97a980004da133ad43/1674545889139/main).
Now the exception is thrown at a different line:
```
java.lang.IllegalArgumentException: Comparison between type 20 and 20 not supported
	at com.arcadedb.serializer.BinaryComparator.compare(BinaryComparator.java:385)
```
Even though my SQL statement does not contain any comparison, the BinaryComparator is invoked by `LSMTreeIndexAbstract.compareKey`, I presume because the engine has to compare the value with the existing values, to determine if it should perform an insert or an update. 
I attach two methods:

- testLocalDateTimeUpsertWithIndex: it shows that, if a DATETIME_MICROS property appears in a unique index, inserting a second row causes java.lang.IllegalArgumentException to be thrown by BinaryComparator
- testLocalDateTimeUpsertWithoutIndex: it shows that if there is no index containing a DATETIME_MICROS property, the second row can be inserted without problems

testLocalDateTimeUpsertWithIndex output:
```
sql String: UPDATE Product SET name = ?, type = ?, start = ?, stop = ?, v = ? UPSERT WHERE name = ?, parameters = [CS_OFFL_STR_ATTREF_20221127T215523_20221129T002322_E001.EEF, STR_ATTREF, 2022-11-27T21:55:23, 2022-11-29T00:23:22, E001, CS_OFFL_STR_ATTREF_20221127T215523_20221129T002322_E001.EEF]
result: { count: 1
 }
Error
com.arcadedb.exception.TransactionException: Transaction error on commit
	at com.arcadedb.database.TransactionContext.commit1stPhase(TransactionContext.java:523)
	at com.arcadedb.database.TransactionContext.commit(TransactionContext.java:114)
	at com.arcadedb.database.EmbeddedDatabase.lambda$commit$2(EmbeddedDatabase.java:366)
	at com.arcadedb.database.EmbeddedDatabase.executeInReadLock(EmbeddedDatabase.java:1326)
	at com.arcadedb.database.EmbeddedDatabase.commit(EmbeddedDatabase.java:361)
	at com.arcadedb.database.EmbeddedDatabase.createRecordNoLock(EmbeddedDatabase.java:800)
	at com.arcadedb.database.EmbeddedDatabase.lambda$createRecord$15(EmbeddedDatabase.java:731)
	at com.arcadedb.database.EmbeddedDatabase.executeInReadLock(EmbeddedDatabase.java:1326)
	at com.arcadedb.database.EmbeddedDatabase.createRecord(EmbeddedDatabase.java:730)
	at com.arcadedb.database.MutableDocument.save(MutableDocument.java:289)
	at com.arcadedb.query.sql.executor.SaveElementStep$1.next(SaveElementStep.java:59)
	at com.arcadedb.query.sql.executor.CountStep.syncPull(CountStep.java:69)
	at com.arcadedb.query.sql.executor.SelectExecutionPlan.fetchNext(SelectExecutionPlan.java:44)
	at com.arcadedb.query.sql.executor.UpdateExecutionPlan.executeInternal(UpdateExecutionPlan.java:62)
	at com.arcadedb.query.sql.parser.UpdateStatement.execute(UpdateStatement.java:116)
	at com.arcadedb.query.sql.parser.Statement.execute(Statement.java:77)
	at com.arcadedb.query.sql.parser.Statement.execute(Statement.java:61)
	at com.arcadedb.query.sql.SQLQueryEngine.command(SQLQueryEngine.java:115)
	at com.arcadedb.database.EmbeddedDatabase.command(EmbeddedDatabase.java:1241)
	at com.arcadedb.server.ServerDatabase.command(ServerDatabase.java:411)
	at eu.serco.csrs.persistence.ArcadeDBTest.testLocalDateTimeComparison(ArcadeDBTest.java:208)
	at eu.serco.csrs.persistence.ArcadeDBTest.main(ArcadeDBTest.java:65)
Caused by: java.lang.IllegalArgumentException: Comparison between type 20 and 20 not supported
	at com.arcadedb.serializer.BinaryComparator.compare(BinaryComparator.java:385)
	at com.arcadedb.index.lsm.LSMTreeIndexAbstract.compareKey(LSMTreeIndexAbstract.java:403)
	at com.arcadedb.index.lsm.LSMTreeIndexMutable.compareKey(LSMTreeIndexMutable.java:232)
	at com.arcadedb.index.lsm.LSMTreeIndexAbstract.lookupInPage(LSMTreeIndexAbstract.java:259)
	at com.arcadedb.index.lsm.LSMTreeIndexAbstract.lookupInPageAndAddInResultset(LSMTreeIndexAbstract.java:509)
	at com.arcadedb.index.lsm.LSMTreeIndexMutable.searchInNonCompactedIndex(LSMTreeIndexMutable.java:395)
	at com.arcadedb.index.lsm.LSMTreeIndexMutable.get(LSMTreeIndexMutable.java:210)
	at com.arcadedb.index.lsm.LSMTreeIndex.lambda$get$7(LSMTreeIndex.java:379)
	at com.arcadedb.utility.RWLockContext.executeInReadLock(RWLockContext.java:52)
	at com.arcadedb.index.lsm.LSMTreeIndex.get(LSMTreeIndex.java:379)
	at com.arcadedb.index.TypeIndex.get(TypeIndex.java:123)
	at com.arcadedb.database.TransactionIndexContext.checkUniqueIndexKeys(TransactionIndexContext.java:323)
	at com.arcadedb.database.TransactionIndexContext.checkUniqueIndexKeys(TransactionIndexContext.java:373)
	at com.arcadedb.database.TransactionIndexContext.commit(TransactionIndexContext.java:162)
	at com.arcadedb.database.TransactionContext.commit1stPhase(TransactionContext.java:479)
	... 21 more
```
testLocalDateTimeUpsertWithoutIndex output:
```
sql String: UPDATE Product SET name = ?, type = ?, start = ?, stop = ?, v = ? UPSERT WHERE name = ?, parameters = [CS_OFFL_STR_ATTREF_20221127T215523_20221129T002322_E001.EEF, STR_ATTREF, 2022-11-27T21:55:23, 2022-11-29T00:23:22, E001, CS_OFFL_STR_ATTREF_20221127T215523_20221129T002322_E001.EEF]
result: { count: 1
 }
sql String: UPDATE Product SET name = ?, type = ?, start = ?, stop = ?, v = ? UPSERT WHERE name = ?, parameters = [CS_OFFL_STR_ATTREF_20220318T215523_20220320T002321_E001.EEF, STR_ATTREF, 2022-03-18T21:55:23, 2022-03-20T00:23:21, E001, CS_OFFL_STR_ATTREF_20220318T215523_20220320T002321_E001.EEF]
result: { count: 1
 }
name = CS_OFFL_STR_ATTREF_20221127T215523_20221129T002322_E001.EEF, type = STR_ATTREF, start = 2022-11-27T21:55:23, stop = 2022-11-29T00:23:22, v = E001
name = CS_OFFL_STR_ATTREF_20220318T215523_20220320T002321_E001.EEF, type = STR_ATTREF, start = 2022-03-18T21:55:23, stop = 2022-03-20T00:23:21, v = E001
```

**Comment by lvca** [comment]
Thanks for the test case, I was able to reproduce. The new types were totally missing from the comparator that as you mentioned is used in the indexes. Just added them and a modified version of your test case. Waiting for all the tests to pass.

**Comment by lvca** [comment]
This is fixed now. Thanks again for the test cases!

**Comment by vic0824** [comment]
Verified with v23.1.2-SNAPSHOT (build 75a93195ad0b26e97ffc7031f361c9b7b51e9065/1674650185130/main)

**Comment by lvca** [comment]
Thanks for confirming.
